### ISSUE TITLE
BertMutation model and optimizer state are shared across repeated runs in the experiment scripts

### ISSUE THREAD

**Issue by WojtAcht** [EC-KitY/BERT-Mutation-GA#2]
Hi, and thanks for releasing the code alongside the paper — it made it easy to follow how the
operator works.

While reading through the experiment runners I noticed that the `BertMutation` instance is
created **outside** the repeated-runs loop, so a single model (and its Adam state) is reused for
all `runs` repetitions of the same instance. I think this may be unintended, since everything else
that carries run state — `evaluator.fitness_dict` — is explicitly cleared inside the loop.

## Where

- `dnm_paper/experiments/artificial_ant.py:46-58` — `build_bert_mutation_operator(...)` is called
  before `for run_index in range(evolution_config.runs)`; only `evaluator.fitness_dict.clear()` is
  inside it.
- `dnm_paper/experiments/frozen_lake.py:47-59` — same structure.

## Reproduction

Against `a569496`, with a small config (3 runs, 2 generations, population 6):

```python
import copy, tempfile, torch
from pathlib import Path
import eckity.algorithms.simple_evolution as se
import dnm_paper.mutation.bert as bertmod
from dnm_paper.config import EvolutionConfig, MutationModelConfig
from dnm_paper.experiments.artificial_ant import run_artificial_ant_experiments

created = []
orig_init = bertmod.BertMutation.__init__
def patched_init(self, *a, **k):
    orig_init(self, *a, **k)
    created.append(self)
bertmod.BertMutation.__init__ = patched_init

snaps = []
orig_evolve = se.SimpleEvolution.evolve
def evolve_spy(self, *a, **k):
    bm = created[-1]
    w = bm.model.bert.embeddings.word_embeddings.weight
    snaps.append((copy.deepcopy(w.detach()), bool(bm.optimizer.state)))
    return orig_evolve(self, *a, **k)
se.SimpleEvolution.evolve = evolve_spy

run_artificial_ant_experiments(
    EvolutionConfig(generations=2, population_size=6, runs=3,
                    instances_to_run={"aux_map1.txt": 30},
                    output_root=Path(tempfile.mkdtemp())),
    MutationModelConfig(batch_size=4, word_embedding_dim=8, internal_size=8,
                        n_layers=1, n_attention_heads=1),
)

print("BertMutation instances created for 3 runs:", len(created))
for i, (w, has_opt_state) in enumerate(snaps):
    print(f"run {i}: weights identical to run 0 start? {torch.equal(w, snaps[0][0])} "
          f"| optimizer has state: {has_opt_state}")
```

Output:

```
BertMutation instances created for 3 runs: 1
run 0: weights identical to run 0 start? True  | optimizer has state: False
run 1: weights identical to run 0 start? False | optimizer has state: True
run 2: weights identical to run 0 start? False | optimizer has state: True
```

So run *k* starts from a model already trained on runs 0…*k*−1 of the same instance.

## Why I think it matters

Each instance is a fixed optimization problem, and `peek_to_best_individual` trains the model to
reproduce genomes from `fitness_dict`. A model carried into run *k* therefore starts with a
learned, instance-specific prior over previously found high-fitness genomes, which it injects into
mutations from generation 0 onward.

The amount of accumulated training differs substantially between the first and last repetition. With
the defaults (`population_size=100`, `mutation_probability=0.4` → ~40 `mutate()` calls per
generation, each logging 2 trajectories since `peek_to_best_individual` also fires, `batch_size=64`):

| Domain | Generations | Trajectories/run | Adam steps/run | Run 0 → run 9 |
|---|---:|---:|---:|---|
| Artificial Ant | 1000 | ~80,000 | ~1,250 | 1,250 → 12,500 |
| Frozen Lake | 100 | ~8,000 | ~125 | 125 → 1,250 |

If the reported curves average over repetitions while the baselines are stateless, this could bias
the comparison, and it also breaks the independence assumption behind the paired permutation test.
I haven't measured the size or even the sign of the effect — a long-lived policy could equally
collapse to low entropy and become *less* effective in later runs — so I don't want to overstate
it. But it seemed worth flagging.

A related detail: the trajectory buffers (`self.rewards`, `self.trajectory_log_probabilities`,
`self.entropy_list`) are only cleared inside `run_epoch()` when a full batch accumulates, so any
partial batch at the end of a run is carried into the first gradient step of the next one.

## Suggested fix

Constructing the operator inside the run loop should be sufficient, e.g. for
`artificial_ant.py`:

```python
evaluator = ArtificialAntEvaluator(instance_matrix)
creator = GAIntVectorCreator(length=individual_length, bounds=(0, 3))

for run_index in range(evolution_config.runs):
    evaluator.fitness_dict.clear()
    mutation_operator = build_bert_mutation_operator(
        mutation_probability=evolution_config.mutation_probability,
        population_size=evolution_config.population_size,
        individual_evaluator=evaluator,
        input_sequence_length=individual_length,
        max_int_val=3,
        mutation_config=mutation_config,
    )
    mutation_operator.mutation_operator.fitness_dict = evaluator.fitness_dict
    ...
```

with the equivalent change in `frozen_lake.py`.

A cheap way to check the impact on already-collected data: regress final fitness (or
area-under-curve) on `run_index` per instance across the existing `results.json` files. A flat
slope would be good evidence that the shared state was inert in practice.

I'm happy to open a PR with the change if that would be useful.

## Environment

- Commit `a569496`
- Python 3.11.5, eckity 0.4.2, torch 2.7.1, transformers 4.50.0

---

*Disclaimer: I only spent a short time with the codebase, and I used Claude Code to help with the
analysis. I did check that the reproduction above runs and produces the quoted output, but I may
well have misread the intent somewhere — please treat this as a question as much as a bug report.*

**Comment by eliadsbgu** [comment]
Thank you very much for raising this issue! 
You’re right, this was a bug. It existed in an older version of my research code and somehow made its way into this repository. It has now been fixed in commit 043c21b. The experiments reported in the current version of the paper were run with the corrected setup, so the reported results are unaffected.
